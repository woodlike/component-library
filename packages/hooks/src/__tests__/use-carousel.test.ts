import { renderHook, act } from '@testing-library/react-hooks';
import { previous, next, useCarousel, jump, move } from '..';

const createTouchEventMock = (x: number, width: number) =>
  (({
    preventDefault: jest.fn(),
    touches: [
      {
        clientX: x,
        target: {
          width,
        },
      },
    ],
  } as unknown) as TouchEvent);

describe('UseCarousel', () => {
  let width: number;
  let length: number;
  beforeEach(() => {
    width = 375;
    length = 6;
  });

  afterEach(() => {
    width = (undefined as unknown) as number;
    length = (undefined as unknown) as number;
  });

  describe('useCarousel()', () => {
    it('should return a default coordinate state of 0', () => {
      const { result } = renderHook(() => useCarousel(4));
      const { coordinate } = result.current;
      expect(coordinate).toBe(0);
    });

    it('should return the previous state coordinate if the current is outsite the length scope', () => {
      const { result } = renderHook(() => useCarousel(4));
      act(() => {
        result.current.jump(2);
      });

      expect(result.current.coordinate).toBe(-50);

      act(() => {
        result.current.jump(5);
      });

      expect(result.current.coordinate).toBe(-50);
    });

    it.each([0, 1, 2, 3, 4, 5, 6, 7, 8])(
      'should return the coordinate of %d',
      current => {
        const { result } = renderHook(() => useCarousel(9));

        act(() => {
          result.current.jump(current);
        });

        expect(result.current.coordinate).toMatchSnapshot();
      },
    );

    it.each([8, 7, 6, 5, 4, 3, 2, 1, 0])(
      'should return the previous coordinate of %d',
      current => {
        const { result } = renderHook(() => useCarousel(9));

        act(() => {
          result.current.jump(current);
          result.current.previous();
        });

        expect(result.current.coordinate).toMatchSnapshot();
      },
    );

    it.each([0, 1, 2, 3, 4, 5, 6, 7, 8])(
      'should return the the next coordinate of %d',
      current => {
        const { result } = renderHook(() => useCarousel(9));

        act(() => {
          result.current.jump(current);
          result.current.next();
        });

        expect(result.current.coordinate).toMatchSnapshot();
      },
    );
  });

  describe('jump()', () => {
    it.each([0, 1, 2, 3, 4])('should jump to %d', current => {
      expect(jump(current, 5)).toMatchSnapshot();
    });
  });

  describe('previous()', () => {
    it.each([10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0])(
      'should return the previous coordinate of %d',
      current => {
        expect(previous(current, 11)).toMatchSnapshot();
      },
    );
  });

  describe('next()', () => {
    it.each([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10])(
      'should return the next coordinate of %d',
      current => {
        expect(next(current, 11)).toMatchSnapshot();
      },
    );
  });

  describe('move()', () => {
    it('should return an absolute 2.666 percentage [swipe right to left]', () => {
      const startEvent = createTouchEventMock(275, width);
      const moveEvent = createTouchEventMock(265, width);
      const moveInit = {
        event: moveEvent,
        startX: startEvent.touches[0].clientX,
        current: 2,
        length,
      };

      expect(move(moveInit)).toMatchObject({ percentage: 2.667 });
    });

    it('should return an absolute 34.667 percentage [swipe left to right]', () => {
      const startEvent = createTouchEventMock(20, width);
      const moveEvent = createTouchEventMock(150, width);
      const moveInit = {
        event: moveEvent,
        startX: startEvent.touches[0].clientX,
        current: 0,
        length,
      };

      expect(move(moveInit)).toMatchObject({ percentage: 34.667 });
    });

    it('should return an absolute 34.667 percentage [swipe right to left]', () => {
      const startEvent = createTouchEventMock(150, width);
      const moveEvent = createTouchEventMock(20, width);
      const moveInit = {
        event: moveEvent,
        startX: startEvent.touches[0].clientX,
        current: 3,
        length,
      };

      expect(move(moveInit)).toMatchObject({ percentage: 34.667 });
    });

    it('should return the coordinate -19.111 [swipe right to left]', () => {
      const startEvent = createTouchEventMock(40, width);
      const moveEvent = createTouchEventMock(-15, width);

      const moveInit = {
        event: moveEvent,
        startX: startEvent.touches[0].clientX,
        current: 1,
        length,
      };
      expect(move(moveInit)).toMatchObject({ coordinate: -19.111 });
    });

    it('it should return the coordinate (-2.444 - currentCoordinate) for each item in the carousel [swipe right to left]', () => {
      const startEvent = createTouchEventMock(40, width);
      const moveEvent = createTouchEventMock(-15, width);

      Array.from({ length }).forEach((_, idx) => {
        const moveInit = {
          event: moveEvent,
          startX: startEvent.touches[0].clientX,
          current: idx,
          length,
        };

        expect(move(moveInit)).toMatchSnapshot();
      });
    });

    it('should return the coordinate -44.667 [swipe left to right]', () => {
      const startEvent = createTouchEventMock(180, width);
      const moveEvent = createTouchEventMock(300, width);
      const moveInit = {
        event: moveEvent,
        startX: startEvent.touches[0].clientX,
        current: 3,
        length,
      };
      expect(move(moveInit)).toMatchObject({ coordinate: -44.667 });
    });

    it('should return the coordinate (-5.333 + currentCoordinate) for each item in the carousel [swipe left to right]', () => {
      const startEvent = createTouchEventMock(180, width);
      const moveEvent = createTouchEventMock(300, width);
      Array.from({ length }).forEach((_, idx) => {
        const moveInit = {
          event: moveEvent,
          startX: startEvent.touches[0].clientX,
          current: idx,
          length,
        };

        expect(move(moveInit)).toMatchSnapshot();
      });
    });
  });
});
