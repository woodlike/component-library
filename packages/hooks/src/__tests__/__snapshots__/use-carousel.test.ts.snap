// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UseCarousel jump() should jump to 0 1`] = `-0`;

exports[`UseCarousel jump() should jump to 1 1`] = `-20`;

exports[`UseCarousel jump() should jump to 2 1`] = `-40`;

exports[`UseCarousel jump() should jump to 3 1`] = `-60`;

exports[`UseCarousel jump() should jump to 4 1`] = `-80`;

exports[`UseCarousel move() it should return the coordinate (-2.444 - currentCoordinate) for each item in the carousel [swipe right to left] 1`] = `
Object {
  "coordinate": -2.444,
  "direction": 1,
  "percentage": 14.667,
}
`;

exports[`UseCarousel move() it should return the coordinate (-2.444 - currentCoordinate) for each item in the carousel [swipe right to left] 2`] = `
Object {
  "coordinate": -19.111,
  "direction": 1,
  "percentage": 14.667,
}
`;

exports[`UseCarousel move() it should return the coordinate (-2.444 - currentCoordinate) for each item in the carousel [swipe right to left] 3`] = `
Object {
  "coordinate": -35.778,
  "direction": 1,
  "percentage": 14.667,
}
`;

exports[`UseCarousel move() it should return the coordinate (-2.444 - currentCoordinate) for each item in the carousel [swipe right to left] 4`] = `
Object {
  "coordinate": -52.444,
  "direction": 1,
  "percentage": 14.667,
}
`;

exports[`UseCarousel move() it should return the coordinate (-2.444 - currentCoordinate) for each item in the carousel [swipe right to left] 5`] = `
Object {
  "coordinate": -69.111,
  "direction": 1,
  "percentage": 14.667,
}
`;

exports[`UseCarousel move() it should return the coordinate (-2.444 - currentCoordinate) for each item in the carousel [swipe right to left] 6`] = `
Object {
  "coordinate": -85.778,
  "direction": 1,
  "percentage": 14.667,
}
`;

exports[`UseCarousel move() should return the coordinate (-5.333 + currentCoordinate) for each item in the carousel [swipe left to right] 1`] = `
Object {
  "coordinate": 5.333,
  "direction": -1,
  "percentage": 32,
}
`;

exports[`UseCarousel move() should return the coordinate (-5.333 + currentCoordinate) for each item in the carousel [swipe left to right] 2`] = `
Object {
  "coordinate": -11.333,
  "direction": -1,
  "percentage": 32,
}
`;

exports[`UseCarousel move() should return the coordinate (-5.333 + currentCoordinate) for each item in the carousel [swipe left to right] 3`] = `
Object {
  "coordinate": -28,
  "direction": -1,
  "percentage": 32,
}
`;

exports[`UseCarousel move() should return the coordinate (-5.333 + currentCoordinate) for each item in the carousel [swipe left to right] 4`] = `
Object {
  "coordinate": -44.667,
  "direction": -1,
  "percentage": 32,
}
`;

exports[`UseCarousel move() should return the coordinate (-5.333 + currentCoordinate) for each item in the carousel [swipe left to right] 5`] = `
Object {
  "coordinate": -61.333,
  "direction": -1,
  "percentage": 32,
}
`;

exports[`UseCarousel move() should return the coordinate (-5.333 + currentCoordinate) for each item in the carousel [swipe left to right] 6`] = `
Object {
  "coordinate": -78,
  "direction": -1,
  "percentage": 32,
}
`;

exports[`UseCarousel next() should return the next coordinate of 0 1`] = `-9.090909090909092`;

exports[`UseCarousel next() should return the next coordinate of 1 1`] = `-18.181818181818183`;

exports[`UseCarousel next() should return the next coordinate of 2 1`] = `-27.272727272727273`;

exports[`UseCarousel next() should return the next coordinate of 3 1`] = `-36.36363636363637`;

exports[`UseCarousel next() should return the next coordinate of 4 1`] = `-45.45454545454545`;

exports[`UseCarousel next() should return the next coordinate of 5 1`] = `-54.54545454545455`;

exports[`UseCarousel next() should return the next coordinate of 6 1`] = `-63.63636363636363`;

exports[`UseCarousel next() should return the next coordinate of 7 1`] = `-72.72727272727273`;

exports[`UseCarousel next() should return the next coordinate of 8 1`] = `-81.81818181818181`;

exports[`UseCarousel next() should return the next coordinate of 9 1`] = `-90.9090909090909`;

exports[`UseCarousel next() should return the next coordinate of 10 1`] = `-0`;

exports[`UseCarousel previous() should return the previous coordinate of 0 1`] = `-90.9090909090909`;

exports[`UseCarousel previous() should return the previous coordinate of 1 1`] = `-0`;

exports[`UseCarousel previous() should return the previous coordinate of 2 1`] = `-9.090909090909092`;

exports[`UseCarousel previous() should return the previous coordinate of 3 1`] = `-18.181818181818183`;

exports[`UseCarousel previous() should return the previous coordinate of 4 1`] = `-27.272727272727273`;

exports[`UseCarousel previous() should return the previous coordinate of 5 1`] = `-36.36363636363637`;

exports[`UseCarousel previous() should return the previous coordinate of 6 1`] = `-45.45454545454545`;

exports[`UseCarousel previous() should return the previous coordinate of 7 1`] = `-54.54545454545455`;

exports[`UseCarousel previous() should return the previous coordinate of 8 1`] = `-63.63636363636363`;

exports[`UseCarousel previous() should return the previous coordinate of 9 1`] = `-72.72727272727273`;

exports[`UseCarousel previous() should return the previous coordinate of 10 1`] = `-81.81818181818181`;

exports[`UseCarousel useCarousel() should return the coordinate of 0 1`] = `-0`;

exports[`UseCarousel useCarousel() should return the coordinate of 1 1`] = `-11.11111111111111`;

exports[`UseCarousel useCarousel() should return the coordinate of 2 1`] = `-22.22222222222222`;

exports[`UseCarousel useCarousel() should return the coordinate of 3 1`] = `-33.333333333333336`;

exports[`UseCarousel useCarousel() should return the coordinate of 4 1`] = `-44.44444444444444`;

exports[`UseCarousel useCarousel() should return the coordinate of 5 1`] = `-55.55555555555556`;

exports[`UseCarousel useCarousel() should return the coordinate of 6 1`] = `-66.66666666666667`;

exports[`UseCarousel useCarousel() should return the coordinate of 7 1`] = `-77.77777777777777`;

exports[`UseCarousel useCarousel() should return the coordinate of 8 1`] = `-88.88888888888889`;

exports[`UseCarousel useCarousel() should return the previous coordinate of 0 1`] = `-88.88888888888889`;

exports[`UseCarousel useCarousel() should return the previous coordinate of 1 1`] = `-0`;

exports[`UseCarousel useCarousel() should return the previous coordinate of 2 1`] = `-11.11111111111111`;

exports[`UseCarousel useCarousel() should return the previous coordinate of 3 1`] = `-22.22222222222222`;

exports[`UseCarousel useCarousel() should return the previous coordinate of 4 1`] = `-33.333333333333336`;

exports[`UseCarousel useCarousel() should return the previous coordinate of 5 1`] = `-44.44444444444444`;

exports[`UseCarousel useCarousel() should return the previous coordinate of 6 1`] = `-55.55555555555556`;

exports[`UseCarousel useCarousel() should return the previous coordinate of 7 1`] = `-66.66666666666667`;

exports[`UseCarousel useCarousel() should return the previous coordinate of 8 1`] = `-77.77777777777777`;

exports[`UseCarousel useCarousel() should return the the next coordinate of 0 1`] = `-11.11111111111111`;

exports[`UseCarousel useCarousel() should return the the next coordinate of 1 1`] = `-22.22222222222222`;

exports[`UseCarousel useCarousel() should return the the next coordinate of 2 1`] = `-33.333333333333336`;

exports[`UseCarousel useCarousel() should return the the next coordinate of 3 1`] = `-44.44444444444444`;

exports[`UseCarousel useCarousel() should return the the next coordinate of 4 1`] = `-55.55555555555556`;

exports[`UseCarousel useCarousel() should return the the next coordinate of 5 1`] = `-66.66666666666667`;

exports[`UseCarousel useCarousel() should return the the next coordinate of 6 1`] = `-77.77777777777777`;

exports[`UseCarousel useCarousel() should return the the next coordinate of 7 1`] = `-88.88888888888889`;

exports[`UseCarousel useCarousel() should return the the next coordinate of 8 1`] = `-0`;
